### DISCUSSION TITLE
Welcome to ztnet Discussions!

### DISCUSSION THREAD

**Discussion by sinamics** [sinamics/ztnet#56] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by mvpdms** [comment]
I'm attempting to install ztnet on a digital ocean machine running docker but I'm not sure what the best course is to properly install it. I'm new to using docker. Just a nudge in the right direction would be much appreciated.

**Comment by sinamics** [comment]
Hi, from the top of my head:
Create a new folder under home named ztnet. 
`mkdir /home/ztnet`

Then copy the content of docker-compose.yml 
https://github.com/sinamics/ztnet#using-docker-compose

Create a new file in DO, 
`nano /home/ztnet/docker-compose.yml`  and paste the content in that file. `Ctrl+x` to save.

Now pull containers and start the app.
`docker-compose up -d`

Application should now be accessable at do_ip:3000.

However, i highly recommend to use a reverse proxy such as nginx with a domain name when accessing ztnet remotely. 
This will give you secure ssl connection if you add letsencrypt certificate.

**Reply by mvpdms** [reply]
Thank you for the reply. I was able to get it up and running on Digital Ocean. Now I just have to figure out how to get my domain name up and running with nginx.

**Reply by mvpdms** [reply]
Got everything up and running beautifully.  Thanks again!

**Comment by tinola** [comment]
> However, i highly recommend to use a reverse proxy such as nginx with a domain name when accessing ztnet remotely.
> This will give you secure ssl connection if you add letsencrypt certificate.

That's absolutely true, but we must constantly keep in mind to block direct WAN access to default port 3000, because nodejs www server is listening wide open on all interfaces via unencrypted HTTP protocol, so there is possibility to compromise access to webUI.
I did this using a firewall rule on my Mikrotik router.

**Reply by tinola** [reply]
To be more precise...
My suggestion is to implement ability to run the NodeJS server on a `lo` interface, for example as a variable in the `.env` file.

Let's say introduce a new variable:
`ZTNET_BIND=`
Default `empty` -> same as now - listen on all (*) NICs

`ZTNET_BIND=127.0.0.1` -> the server listens only on the `localhost` interface

In this case, we can safely run a reverse proxy on HTTPS with a letsencrypt certificate without configuring the firewall rules.

**Comment by sinamics** [comment]
@tinola 
You can bind the application to the loopback interface using the ip:port in docker-compose
```yml
  ztnet:
    ports:
      - 127.0.0.1:3000:3000   
```

Then use reverse proxy to point to the lo interface.

**Reply by tinola** [reply]
I would love to help you with this, but as you probably remember, I use **FreeBSD** system without `docker` support, so in my case (and probably a lot of other non-Linux users) `ztnet` is working as a standalone application...

I don't know if it's possible, but if we add a new variable to the `.env` file, let's say:

`ZTNET_BIND_IFACE="127.0.0.1"`

then the `server.js` would only listen on the **loopback** interface, instead of **all** interfaces **(`*`)**.

In my case... I have got `ztnet` WWW "core" unencrypted server running on port 4000 (set by env `PORT` variable in `~/.cshrc` file).
Let's run FreeBSD's command `sockstat`:

```
# sockstat -4 | grep 4000
ztnet   node       78949 16 tcp46  *:4000                *:*
```

Now: `*:4000` is crucial... asterisk (`*`) before port means "listen on all interfaces", but we want to listen on **127.0.0.1 (lo)** only...

That's why I created firewall rule to secure webUI from WAN :)

**Comment by mvpdms** [comment]
What is the best/easiest way to update ztnet if its running on docker?

**Reply by sinamics** [reply]
if you have `image: sinamics/ztnet:latest` in you docker-compose.yml file, then simply run these commands:
```bash
docker-compose pull
docker-compose up -d
````

**Reply by mvpdms** [reply]
I updated the file and it confirmed the update of zerotier and ztnet. I'm still getting the all white background with white text when I initially go to the login. Any suggestions? 

<img width="748" alt="Screenshot_1" src="https://github.com/sinamics/ztnet/assets/32503385/d3b9fdc3-2d60-4067-8374-a54452dd4574">

**Reply by sinamics** [reply]
Its probably a cache issue. Try to hit shift + f5, or try icognito mode

**Comment by mvpdms** [comment]
When I go incognito is when I see the issue. I was trying to replicate a
new user going to the site.

**Reply by sinamics** [reply]
and you are 100% sure you actually using the new version, 0.4.0 ?

**Reply by mvpdms** [reply]
It upgraded to 0.4.0 and is working now. As always thank you for the help!

**Comment by wmd1942** [comment]
I got it running within a few min on the Ubuntu container inside my Promox. Thanks a lot for Sinamics' fantastic work.
